### PR TITLE
Add to_erl/from_erl conversions to Time/Date/NaiveDateTime

### PR DESCRIPTION
**Pull request by bitwalker**
Implements #4671

I chose to use `{:error, :unsupported_calendar}` for the error tuple in cases where `Calendar.ISO` is not the calendar for the given `Date` or `NaiveDateTime`, I can change that if there is a more descriptive atom to use.

Been awhile since my last PR, so let me have it if I've forgotten something :)

### CODE REVIEW

**Comment by josevalim** [comment]
> I chose to use {:error, :unsupported_calendar} for the error tuple in cases where Calendar.ISO is not the calendar for the given Date or NaiveDateTime, I can change that if there is a more descriptive atom to use.

Let's raise for now because this is committing to an API and we have no idea if unsupported_calendar is what we would like to return in the future. :)

File: lib/elixir/lib/calendar.ex [hunk 1]
```diff
@@ -92,6 +92,42 @@ defmodule Date do
...
+  Attempting to convert an invalid ISO calendar date will produce an error tuple.
+
+  ## Examples
+
+    iex> Date.from_erl({2000, 1, 1})
+    {:ok, %Date{calendar: Calendar.ISO, year: 2000, month: 1, day: 1}}
+    iex> Date.from_erl({2000, 13, 1})
+    {:error, :invalid_date}
+  """
+  @spec from_erl(:calendar.date()) :: {:ok, __MODULE__.t} | {:error, atom}
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Use `Date.t` for consistency. :)

File: lib/elixir/lib/calendar.ex [hunk 2]
```diff
@@ -137,6 +173,36 @@ defmodule Time do
...
+  Converts an Erlang time tuple (like those produced by the `:calendar` module) to a `Time` struct.
+
+  ## Examples
+
+    iex> Time.from_erl({23, 30, 15})
+    {:ok, %Time{hour: 23, minute: 30, second: 15, microsecond: 0}}
+    iex> Time.from_erl({24, 30, 15})
+    {:error, :invalid_time}
+  """
+  @spec from_erl(:calendar.time()) :: {:ok, __MODULE__.t} | {:error, atom}
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: `Time.t`

File: lib/elixir/lib/calendar.ex [hunk 3]
```diff
@@ -92,6 +92,42 @@ defmodule Date do
...
+  def to_erl(%__MODULE__{}), do: {:error, :unsupported_calendar}
+
+  @doc """
+  Converts an Erlang date tuple (like those produced by the `:calendar` module) to a `Date` struct.
+
+  Attempting to convert an invalid ISO calendar date will produce an error tuple.
+
+  ## Examples
+
+    iex> Date.from_erl({2000, 1, 1})
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Code samples must be four spaces.

File: lib/elixir/lib/calendar.ex [hunk 4]
```diff
@@ -137,6 +173,36 @@ defmodule Time do
...
+  def to_erl(%__MODULE__{} = time) do
+    {:ok, {time.hour, time.minute, time.second}}
+  end
+
+  @doc """
+  Converts an Erlang time tuple (like those produced by the `:calendar` module) to a `Time` struct.
+
+  ## Examples
+
+    iex> Time.from_erl({23, 30, 15})
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Code samples must be four spaces.

File: lib/elixir/lib/calendar.ex [hunk 5]
```diff
@@ -207,6 +273,46 @@ defmodule NaiveDateTime do
...
+  def to_erl(%__MODULE__{}), do: {:error, :unsupported_calendar}
+
+  @doc """
+  Converts an Erlang datetime tuple (like those produced by the `:calendar` module) to a `NaiveDateTime` struct.
+
+  Attempting to convert an invalid ISO calendar date will produce an error tuple.
+
+  ## Examples
+
+    iex> NaiveDateTime.from_erl({{2000,1,1},{13,30,15}})
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Code samples must be four spaces.

**Comment by josevalim** [comment]
Actually, ignore my previous comment. from_erl and to_erl will only work with ISO because the Erlang format only works with ISO.

File: lib/elixir/lib/calendar.ex [hunk 6]
```diff
@@ -92,6 +92,42 @@ defmodule Date do
       {:error, :invalid_date}
     end
   end
+
+  @doc """
+  Converts a `Date` struct to an Erlang date tuple (like those produced by the `:calendar` module).
+
+  Only supports converting Dates which are in the ISO calendar,
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Dates should be lowercase or be `Date`s. The same for Times and NaiveDateTimes below.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by bitwalker** [comment]
Thanks for the feedback, comments noted for future reference ;)

**Pull request merged**